### ISSUE TITLE
ryu patching is understood and fixed for 1.8

### ISSUE THREAD

**Issue by nstng** [magma/magma#13687]
### Problem
ryu is not patched with magma specific QFI changes in master and the 1.8 release

#### What patches are there?
* `lte/gateway/deploy/roles/magma/files/patches` contains four patches
* the patch `lte/gateway/deploy/roles/magma/files/patches/ryu_ipfix_args.patch` seems to be applied to the checked-in files `lte/gateway/deploy/roles/magma/files/nx_actions.py` and `lte/gateway/deploy/roles/magma/files/nx_actions_3.5.py` for python `3.8` and `3.5`.
  * note: there is also `cwf/gateway/deploy/roles/ovs/files/nx_actions.py` which seems to differ from the file above
* in the same way the patch `lte/gateway/deploy/roles/magma/files/patches/0001-Set-unknown-dpid-ofctl-log-to-debug.patch` is pre-applied to `lte/gateway/deploy/roles/magma/files/service.py`.
* note: multiple code formattings were applied to the pre-patched files - this makes it hard to see what patch was actually applied
  * verified manually that `lte/gateway/deploy/roles/magma/files/nx_actions.py` without any formatting changes coincides to a freshly patched `nx_actions.py` from the official 4.34 repo.
* the other patches (`0002-`-`0003-`) are not pre-applied to any files.

##### for the vm - provisioning
* for the magma dev VM ryu is patch during provisioning - see `lte/gateway/deploy/roles/magma/tasks/main.yml:424`+
* ryu seems not to be installed during provisioning for a long time, i.e., patching does not really make sense here
* additionally `ryu_ipfix_args.patch` is applied by overriding `nx_actions.py` by the pre-applied patched files (see above) where `3.5` and `3.8` seems to be mixed up
* the other patches are not applied - the patch files would overriding ryu files (if ryu was installed) and by this damaging ryu

##### for the vm - build
* see `orc8r/gateway/python/python.mk:56`+
* triggered via `make run`
* this is why the patches are recognized in the integration tests
* note: `0001-Set-unknown-dpid-ofctl-log-to-debug.patch` is applied to the wrong file and is failing

##### via bazel
* see `bazel/python_repositories.bzl`
* this is not yet used in production builds

##### for magma.deb
* `nx_actions_3.5.py` ans `service.py` (see above) are deployed (but not `nx_actions.py`) and they override the respective files in ryu by `lte/gateway/release/magma-postinst`.
* This is, ryu is installed without patches and the post-install script applies (two of the four) patches by overriding ryu files by pre-patched files.

##### for containerized AGW
* done in `lte/gateway/docker/services/python/Dockerfile:71` via `make buildenv`
* relevant patches are applied via `orc8r/gateway/python/python.mk:56` (not on `service.py` but this might be not that relevant)
* checked with @jheidbrink and @ajahl 

#### Recommendations for moving forward
* for 1.8
  * manually build a patched ryu package and add it to the magma artifactory
    * make sure this is not replaced by a native ubuntu package
* not for 1.8
  * remove the ansible patching - this is just damaged
  * remove the pre-patched files and use a patched ryu .deb package or apply the patches during a build

**Comment by nstng** [comment]
draft for a script to build a patched ryu package:
```
on magma dev VM

mkdir /tmp/magma-ryu
cd magma-ryu

git clone --depth 1 --branch v4.34 git@github.com:faucetsdn/ryu.git

PATCHES_DIR=$MAGMA_ROOT/lte/gateway/deploy/roles/magma/files/patches

patch --dry-run -N -s -f ryu/ryu/ofproto/nx_actions.py <$PATCHES_DIR/ryu_ipfix_args.patch 2>/dev/null &&  (patch -N -s -f ryu/ryu/ofproto/nx_actions.py <$PATCHES_DIR/ryu_ipfix_args.patch && echo "ryu was patched" ) || ( true && echo "skipping ryu patch since it was already applied")

patch --dry-run -N -s -f ryu/ryu/ofproto/nx_actions.py <$PATCHES_DIR/0001-Set-unknown-dpid-ofctl-log-to-debug.patch 2>/dev/null &&  (patch -N -s -f ryu/ryu/ofproto/nx_actions.py <$PATCHES_DIR/0001-Set-unknown-dpid-ofctl-log-to-debug.patch && echo "ryu was patched" ) || ( true && echo "skipping ryu patch since it was already applied")

patch --dry-run -N -s -f ryu/ryu/ofproto/nicira_ext.py <$PATCHES_DIR/0002-QFI-value-set-in-Openflow-controller-using-RYU.patch 2>/dev/null &&  (patch -N -s -f ryu/ryu/ofproto/nicira_ext.py <$PATCHES_DIR/0002-QFI-value-set-in-Openflow-controller-using-RYU.patch && echo "ryu was patched" ) || ( true && echo "skipping ryu patch since it was already applied")

patch --dry-run -N -s -f ryu/ryu/ofproto/nx_match.py <$PATCHES_DIR/0003-QFI-value-set-in-Openflow-controller-using-RYU.patch 2>/dev/null &&  (patch -N -s -f ryu/ryu/ofproto/nx_match.py <$PATCHES_DIR/0003-QFI-value-set-in-Openflow-controller-using-RYU.patch && echo "ryu was patched" ) || ( true && echo "skipping ryu patch since it was already applied")

fpm -s python -t deb --no-python-dependencies --python-package-name-prefix=python3 --python-bin=python3 ryu/setup.py
```

**Comment by nstng** [comment]
#13705 is an approach to clean-up the ryu patching. This is **NOT** relevant for 1.8!

**Comment by prabinakpattnaik** [comment]
@nstng , I have verified with above RYU script in my local setup. It's working fine.
Created PDU session with QFI value.
```
sudo ovs-ofctl -O OpenFlow13 dump-flows gtp_br0 table=0
sudo: unable to resolve host prabina-qemu-agw: Name or service not known
 cookie=0x0, duration=416.982s, table=0, n_packets=0, n_bytes=0, priority=65503,tun_id=0x7fffffff,qfi=8,in_port=32768 actions=set_field:02:00:00:00:00:01->eth_src,set_field:ff:ff:ff:ff:ff:ff->eth_dst,set_field:0x2710->reg9,set_field:0x181c9->metadata,resubmit(,1)
 cookie=0x0, duration=416.954s, table=0, n_packets=0, n_bytes=0, priority=65503,ip,in_port=LOCAL,nw_dst=192.168.128.11 actions=set_field:0x2710->tun_id,set_field:172.16.5.207->tun_dst,set_field:0x8000->reg8,set_field:oam->tun_flags,set_field:0x8->qfi,set_field:0x181c9->metadata,resubmit(,1)
 cookie=0x0, duration=416.954s, table=0, n_packets=0, n_bytes=0, priority=65503,ip,in_port=mtr0,nw_dst=192.168.128.11 actions=set_field:0x2710->tun_id,set_field:172.16.5.207->tun_dst,set_field:0x8000->reg8,set_field:oam->tun_flags,set_field:0x8->qfi,set_field:0x181c9->metadata,resubmit(,1)
 cookie=0x0, duration=416.954s, table=0, n_packets=0, n_bytes=0, priority=65503,arp,in_port=LOCAL,arp_tpa=192.168.128.11 actions=set_field:0x181c9->metadata,resubmit(,1)
 cookie=0x0, duration=416.954s, table=0, n_packets=0, n_bytes=0, priority=65503,arp,in_port=mtr0,arp_tpa=192.168.128.11 actions=set_field:0x181c9->metadata,resubmit(,1)
 cookie=0x0, duration=2302.803s, table=0, n_packets=2, n_bytes=140, priority=10,in_port=ipfix0 actions=goto_table:201
 cookie=0x0, duration=2302.803s, table=0, n_packets=24, n_bytes=1176, priority=0 actions=goto_table:1
```

**Comment by nstng** [comment]
Uploaded patched ryu 4.34 to focal-ci and focal-1.8.0. In pairing with @mpfirrmann .
